ethstorage/scanner/worker.go

```diff
@@ -0,0 +1,214 @@
+				err = es.ErrCommitMismatch
+			}
+		} else if s.cfg.Mode == modeCheckBlob {
+			// query blob and check meta from storage
+			_, found, err = s.sm.TryRead(kvIndex, int(s.sm.MaxKvSize()), commit)
+		} else {
+			panic(fmt.Sprintf("invalid scanner mode: %d", s.cfg.Mode))
+		}
+		if found && err == nil {
+			// happy path
+			s.lg.Debug("Scanner: check KV done", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if !found {
+			// the shard does not exist locally
+			s.lg.Warn("Scanner: blob not found", "kvIndex", kvIndex, "commit", commit)
+			continue
+		}
+		if err != nil {
+			// chances are the blob is downloaded but not written to the storage
+			if blob := s.loadKvFromCache(kvIndex, commit); blob != nil {
+				s.lg.Debug("Scanner: check KV done from cache", "kvIndex", kvIndex, "commit", commit)
+				continue
+			}
+			s.lg.Error("Scanner: read blob error", "kvIndex", kvIndex, "commit", commit.Hex(), "err", err)
+			if err == es.ErrCommitMismatch {
+				sts.mismatched = append(sts.mismatched, kvIndex)
+				if err := s.fixKv(kvIndex, commit); err != nil {
+					sts.failed = append(sts.failed, kvIndex)
+					s.lg.Error("Scanner: fix blob error", "kvIndex", kvIndex, "err", err)
+					sendError(kvIndex, fmt.Errorf("failed to fix blob: %w", err))
+				} else {
+					sts.fixed = append(sts.fixed, kvIndex)
+					// remove the error from the cache
+					sendError(kvIndex, nil)
+				}
+			} else {
+				sendError(kvIndex, fmt.Errorf("failed to read blob: %w", err))
+			}
+		}
+	}
+	s.nextKvIndex = nextKvIndex
+	if len(kvsInBatch) > 0 {
+		s.lg.Info("Scanner: scan batch done", "from", kvsInBatch[0], "to", kvsInBatch[len(kvsInBatch)-1], "count", len(kvsInBatch), "nextKvIndex", nextKvIndex)
+	}
+	return &sts, nil
+}
+
+func (s *Worker) determineBatchIndexRange(localKvs []uint64) ([]uint64, uint64, error) {
+	localKvTotal := uint64(len(localKvs))
+
+	batchStart := s.nextKvIndex
+	if batchStart >= localKvTotal {
+		batchStart = 0
+		s.lg.Info("Scanner: scan batch start over")
+	}
+	batchEnd := batchStart + uint64(s.cfg.BatchSize)
+	if batchEnd > localKvTotal {
+		batchEnd = localKvTotal
+	}
+	return localKvs[batchStart:batchEnd], batchEnd, nil
+}
+
+func (s *Worker) queryLocalKvs() ([]uint64, error) {
+	kvEntryCount, err := s.l1.GetStorageLastBlobIdx(rpc.LatestBlockNumber.Int64())
+	if err != nil {
+		return nil, fmt.Errorf("failed to query total KV entries: %w", err)
+	}
+	s.lg.Info("Scanner: query KV entry count done", "kvEntryCount", kvEntryCount)
+	var localKvs []uint64
+	kvEntries := s.sm.KvEntries()
+	localShards := s.sm.Shards()
+	s.lg.Info("Scanner: local shards", "shards", localShards)
+	sort.Slice(localShards, func(i, j int) bool {
+		return localShards[i] < localShards[j]
+	})
+	for _, shardId := range localShards {
+		for k := uint64(0); k < kvEntries; k++ {
+			kvIdx := shardId*kvEntries + k
+			if kvIdx < kvEntryCount {
+				// only check non-empty data
+				localKvs = append(localKvs, kvIdx)
```

Comment:
This approach can lead to high memory usage when dealing with large shards. For instance, a 2 TB shard would consume around 122 MB of memory, which may become a bottleneck depending on the node’s resource constraints.